### PR TITLE
add floating point exceptions and detect invalid material model usage

### PR DESCRIPTION
**Pull request by tjhei**
- enable exception handling for floating point operations involving NaNs 
- fix a bunch of floating point bugs
- set material model inputs to NaN

### CODE REVIEW

**Comment by tjhei** [comment]
not quite ready (a bunch of test are still broken)

File: source/main.cc [hunk 1]
```diff
@@ -375,6 +377,9 @@ int main (int argc, char *argv[])
   // as deemed useful by TBB
   Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, /*n_threads =*/ 1);
 
+  // enable floating point exceptions
+  feenableexcept(FE_DIVBYZERO|FE_INVALID);
```
Comment (bangerth) [inline source/main.cc]: is this expensive?

Comment (bangerth) [inline source/main.cc]: I mean, is it expensive to run with this enabled? We could do it only if debug mode is enabled.

**Comment by bangerth** [comment]
Interesting approach. I like it, but wonder whether execution is slower with this enabled?

**Comment by tjhei** [comment]
Good point. I will wrap it in #ifdef DEBUG.

**Comment by tjhei** [comment]
I checked the test outputs and addressed all comments. I hope the tests all pass. If yes, we are ready to go.

**Comment by bangerth** [comment]
OK once tests pass.

**Comment by bangerth** [comment]
These tests currently fail:
```

The following tests FAILED:
	 54 - depth-dependent-3d-shell-list-simple (Failed)
	 55 - depth-dependent-box-file-simple (Failed)
	 56 - depth-dependent-box-function-simple (Failed)
	 57 - depth-dependent-box-none-simple (Failed)
	 64 - diffusion_dislocation (Failed)
```
The last one is probably due to your branch point and was fixed last night. The others look worth looking into.

**Comment by bangerth** [comment]
Indeed, these are all floating point exceptions.

**Comment by tjhei** [comment]
okay, tests are passing now. Care to review the last couple of changes?

**Pull request merged**